File: problems/0239_sliding-window-maximum/solution_heap_lazy.java
import java.util.*;

class Solution {

    public int[] maxSlidingWindow(int[] nums, int k) {
        // Max-heap of (value, index) records, ordered by value.
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> Integer.compare(b[0], a[0]));
        int[] result = new int[Math.max(0, nums.length - k + 1)];
        int ri = 0;
        for (int i = 0; i < nums.length; i++) {
            heap.offer(new int[] { nums[i], i });
            // Lazy deletion: drop records whose index has slid out of the window.
            while (heap.peek()[1] <= i - k) heap.poll();
            // The top is now the largest value still inside the window.
            if (i >= k - 1) result[ri++] = heap.peek()[0];
        }
        return result;
    }
}

File: problems-originals/0201-0300/0239_sliding-window-maximum/solution_heap_lazy.cpp
class Solution {
  public:
    vector<int> maxSlidingWindow(vector<int> &nums, int k) {
        int n = nums.size();
        // Max-heap of (value, index) records; pair ordering compares value first.
        priority_queue<pair<int, int>> heap;
        vector<int> result;
        result.reserve(n >= k ? n - k + 1 : 0);
        for (int i = 0; i < n; ++i) {
            heap.push({nums[i], i});
            // Lazy deletion: pop records whose index has slid out of the window.
            while (heap.top().second <= i - k)
                heap.pop();
            // The top is now the largest value still inside the window.
            if (i >= k - 1)
                result.push_back(heap.top().first);
        }
        return result;
    }
};
